server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetadataUpdater.java

```diff
@@ -167,10 +167,12 @@ private static IndexMetadata updateInSyncAllocations(
                 updatedIndexMetadata = updatedIndexMetadata.withInSyncAllocationIds(shardId.id(), Set.of());
             } else {
                 final String allocationId;
+
                 if (recoverySource == RecoverySource.ExistingStoreRecoverySource.FORCE_STALE_PRIMARY_INSTANCE) {
                     allocationId = RecoverySource.ExistingStoreRecoverySource.FORCED_ALLOCATION_ID;
-                    updatedIndexMetadata = updatedIndexMetadata.withTimestampRange(
-                        updatedIndexMetadata.getTimestampRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards())
+                    updatedIndexMetadata = updatedIndexMetadata.withTimestampRanges(
+                        updatedIndexMetadata.getTimestampRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards()),
+                        updatedIndexMetadata.getEventIngestedRange().removeShard(shardId.id(), oldIndexMetadata.getNumberOfShards())
```

Comment:
Thanks for flagging this. It's a part of the code that makes me very nervous, as I don't really understand what is happening here.

I did ask David Turner about it in an earlier version of the PR: https://github.com/elastic/elasticsearch/pull/106252/files/126eb4ddf091079fbeba09654409bddac78f822e#r1569022493

You are right that `removeShard` can return NO_SHARDS under either of the following conditions:

```
        if (shards == null && numberOfShards == 1) {
```
or
```
        if (shards != null && shards.length == 1 && shards[0] == shardId) {
```

I don't really understand what is going on when we call removeShard here. The comments indicate that this is checking "if we have been force-initializing an empty primary or a stale primary" and if so should remove a shard from the range. 

> I think this would result in the event ingested timestamp range becoming NO_SHARDS and thus break assertions elsewhere?

So when does reverting to NO_SHARDS break things? What is the right way to handle this? Perhaps if eventIngestedRange is already UNKNOWN, we should have an if-guard here to make sure it doesn't revert to NO_SHARDS?

But then we'd have different handling between @timestamp and event.ingested, as [this test](https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/index/shard/IndexLongFieldRangeTests.java#L104) shows that UNKNOWN can convert to NO_SHARDS under some scenarios.

Reply:
Yeah, going from UNKNOWN to NO_SHARDS is fine when the minimum transport version check allows it. The problem is the missing check for that, not the code otherwise.

I wonder if we should let `withTimeStampRanges` have a transport version and thus centralize the minimum version check there. That could just disregard the ingested timestamp if the check does not pass. And it would make it more obvious that we have caught all cases of manipulating this correctly.

Reply:
OK, thanks. I've tried that in [this commit](https://github.com/elastic/elasticsearch/commit/0e69ac3c7785499cbf3de6977d2c5f414aaebe9b). It required changing several method signatures along the chain.

Reply:
I'll leave a PR review comment around this. Overall looks ok but I think we can remove at least one of the signature changes.